### ISSUE TITLE
False positive `unused-argument` in `dataclass` `__new__`

### ISSUE THREAD

**Issue by michele-riva** [pylint-dev/pylint#9843]
### Bug description

Looks like #3670 is not quite solved for classes that dynamically define an `__init__`. While this is probably complex to handle in the general case, `dataclasses` are a common use case that could be handled in a specific manner. Here a minimal example:

```python
"""File unused_arg_dataclass.py for testing unused-argument on __new__ for dataclasses."""

from dataclasses import dataclass

@dataclass
class Data:  # pylint: disable=missing-class-docstring
    arg: ...

    def __new__(cls, *args, **kwargs):   # unused-argument for args, kwargs
        instance = super().__new__(cls)
        # ... do stuff with instance without using args, kwargs...
        return instance
```

The current solution for not emitting `unused-argument` for `__new__` (added in #8542) is at https://github.com/pylint-dev/pylint/blob/c25bef3ab28d721ace050b01ba5ab7b3a34d60fb/pylint/checkers/variables.py#L2735-L2747

### Command used

```shell
pylint unused_arg_dataclass.py
```

### Pylint output

```python
************* Module unused_arg_dataclass
unused_arg_dataclass.py:9:0: W0613: Unused argument 'args' (unused-argument)
unused_arg_dataclass.py:9:0: W0613: Unused argument 'kwargs' (unused-argument)
```

### Expected behavior

No `unused-argument` for `dataclass` `__new__`.

### Pylint version

```shell
pylint 3.2.6
astroid 3.2.4
Python 3.12.0
```
